rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/server/RMServiceImpl.java

```diff
@@ -519,201 +359,163 @@ public String releaseNode(String sessionId, String url) throws RestServerExcepti
-                default:
-                    throw new RestServerException(code, ret);
-            }
-        } catch (IOException e) {
-            throw new ServiceException("Failed to read server response", e);
-        } finally {
-            if (clientResponse != null) {
-                clientResponse.releaseConnection();
+    @Override
+    public String getStatHistory(final String sessionId, final String range) throws RestServerException, ServiceException {
+        return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
+            @Override
+            public InputStream apply(RestClient restClient) {
+                return restClient.getStatHistory(sessionId, range);
             }
-        }
+        });
     }
 
     @Override
-    public String getStatHistory(String sessionId, String range) throws RestServerException, ServiceException {
-        RestClient client = ProxyFactory.create(RestClient.class, RMConfig.get().getRestUrl(), executor);
-        ClientResponse<InputStream> clientResponse = null;
-        try {
-            clientResponse = client.getStatHistory(sessionId, range);
-            int code = clientResponse.getStatus();
-            String ret = convertToString(clientResponse.getEntity());
-
-            switch (code) {
-                case 200:
-                    return ret;
-                default:
-                    throw new RestServerException(code, ret);
+    public String executeNodeScript(final String sessionId, final String script, final String engine, final String nodeUrl) throws RestServerException, ServiceException {
+        return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
+            @Override
+            public InputStream apply(RestClient restClient) {
+                return restClient.executeNodeScript(sessionId, nodeUrl, script, engine);
             }
-        } catch (IOException e) {
-            throw new ServiceException("Failed to read server response", e);
-        } finally {
-            if (clientResponse != null) {
-                clientResponse.releaseConnection();
+        });
+    }
+
+    private RestClient getRestClientProxy() {
+        ResteasyClient client = new ResteasyClientBuilder().asyncExecutor(threadPool).build();
+        ResteasyWebTarget target = client.target(RMConfig.get().getRestUrl());
+
+        return target.proxy(RestClient.class);
+    }
+
+    private String executeFunction(BiFunction<RestClient, Set<String>, InputStream> action, Set<String> urls,
+                                   String actionName) throws ServiceException, RestServerException {
+
+        RestClient restClientProxy = getRestClientProxy();
+
+        int failures = 0;
+        int success = 0;
+
+        for (String url : urls) {
+            InputStream inputStream = null;
+
+            try {
+                inputStream = action.apply(restClientProxy, Collections.singleton(url));
```

Comment:
Do we use the inputStream? if not is it possible to use a Consumer instead of a Function?